system/security/plugins/testauthSecurity/testauthSecurity.cpp

```diff
@@ -0,0 +1,591 @@
+    }
+
+public:
+    CTestAuthSecurityManager(const char* serviceName, IPropertyTree* secMgrCfg, IPropertyTree* bindConfig)
+        : CBaseSecurityManager(serviceName, (IPropertyTree* ) nullptr)
+    {
+        if (!secMgrCfg || !bindConfig)
+            return;
+        readUserAccess(secMgrCfg, bindConfig);
+    }
+
+    SecFeatureSet queryFeatures(SecFeatureSupportLevel level) const override
+    {
+        SecFeatureSet mgrFeatures = CBaseSecurityManager::queryFeatures(level);
+        switch (level)
+        {
+        case SFSL_Safe:
+            mgrFeatures = mgrFeatures | s_safeFeatureMask;
+            break;
+        case SFSL_Implemented:
+            mgrFeatures = mgrFeatures | s_implementedFeatureMask;
+            break;
+        case SFSL_Unsafe:
+            mgrFeatures = mgrFeatures & ~s_safeFeatureMask;
+            break;
+        default:
+            break;
+        }
+        return mgrFeatures;
+    }
+
+    secManagerType querySecMgrType() override {	return SMT_TestAuth; }
+    inline virtual const char* querySecMgrTypeName() override { return "testauth"; }
+    const char* getDescription() override { return "Test Auth Security Manager"; }
+
+    IAuthMap* createAuthMap(IPropertyTree* authconfig, IEspSecureContext* secureContext = nullptr) override
+    {
+        Owned<CAuthMap> authmap = new CAuthMap();
+
+        Owned<IPropertyTreeIterator> iter = authconfig->getElements(".//Location");
+        ForEach(*iter)
+        {
+            IPropertyTree& location = iter->query();
+            StringBuffer pathStr, rstr, required, description;
+            location.getProp("@path", pathStr);
+            location.getProp("@resource", rstr);
+            location.getProp("@required", required);
+            location.getProp("@description", description);
+
+            if (pathStr.length() == 0)
+                throw makeStringException(-1, "path empty in Authenticate/Location");
+            if (rstr.length() == 0)
+                throw makeStringException(-1, "resource empty in Authenticate/Location");
+
+            ISecResourceList* rlist = authmap->queryResourceList(pathStr);
+            if (rlist == nullptr)
+            {
+                rlist = createResourceList("htpasswdsecurity", secureContext);
+                authmap->add(pathStr, rlist);
+            }
+            ISecResource* rs = rlist->addResource(rstr);
+            SecAccessFlags requiredaccess = str2perm(required);
+            rs->setRequiredAccessFlags(requiredaccess);
+            rs->setDescription(description);
+        }
+        return authmap.getClear();
+    }
+
+    IAuthMap* createFeatureMap(IPropertyTree* authconfig, IEspSecureContext* secureContext = nullptr) override
+    {
+        Owned<CAuthMap> authmap = new CAuthMap();
+
+        Owned<IPropertyTreeIterator> iter = authconfig->getElements(".//Feature");
+        ForEach(*iter)
+        {
+            IPropertyTree& feature = iter->query();
```

Comment:
similar comment about shared code in the base class.

Reply:
Will do